Firestore/core/src/firebase/firestore/immutable/array_sorted_map.h

```diff
@@ -0,0 +1,321 @@
+    size_type new_size = size_ + appending;
+    assert(new_size <= fixed_size);
+
+    std::copy(src_begin, src_end, end());
+    size_ = new_size;
+  }
+
+  /**
+   * Appends a single value to the array.
+   */
+  void append(T&& value) {
+    size_type new_size = size_ + 1;
+    assert(new_size <= fixed_size);
+
+    *end() = std::move(value);
+    size_ = new_size;
+  }
+
+  const_iterator begin() const {
+    return contents_.begin();
+  }
+
+  const_iterator end() const {
+    return begin() + size_;
+  }
+
+  size_type size() const {
+    return size_;
+  }
+
+ private:
+  iterator begin() {
+    return contents_.begin();
+  }
+
+  iterator end() {
+    return begin() + size_;
+  }
+
+  array_type contents_;
+  size_type size_;
+};
+
+}  // namespace impl
+
+/**
+ * ArraySortedMap is a value type containing a map. It is immutable, but has
+ * methods to efficiently create new maps that are mutations of it.
+ */
+template <typename K, typename V, typename C = std::less<K>>
+class ArraySortedMap : public impl::ArraySortedMapBase {
+ public:
+  using this_type = ArraySortedMap<K, V, C>;
+  using key_comparator_type = KeyComparator<K, V, C>;
+
+  /**
+   * The type of the entries stored in the map.
+   */
+  using value_type = std::pair<K, V>;
+
+  /**
+   * The type of the fixed-size array containing entries of value_type.
+   */
+  using array_type = impl::FixedArray<value_type, kFixedSize>;
+  using const_iterator = typename array_type::const_iterator;
+
+  using array_pointer = std::shared_ptr<const array_type>;
+
+  /**
+   * Creates an empty ArraySortedMap.
+   */
+  explicit ArraySortedMap(const C& comparator = C())
+      : array_(EmptyArray()), key_comparator_(comparator) {
+  }
+
+  /**
+   * Creates an ArraySortedMap containing the given entries.
+   */
+  ArraySortedMap(std::initializer_list<value_type> entries,
+                 const C& comparator = C())
+      : array_(std::make_shared<array_type>(entries.begin(), entries.end())),
+        key_comparator_(comparator) {
+  }
+
+  /**
+   * Creates a new map identical to this one, but with a key-value pair added or
+   * updated.
+   *
+   * @param key The key to insert/update.
+   * @param value The value to associate with the key.
+   * @return A new dictionary with the added/updated value.
+   */
+  this_type insert(const K& key, const V& value) const {
+    const_iterator current_end = end();
+    const_iterator pos = LowerBound(key);
+    bool replacing_entry = false;
+
+    if (pos != current_end) {
+      // LowerBound found an entry where pos->first >= pair.first. Reversing the
+      // argument order here tests pair.first < pos->first.
+      replacing_entry = !key_comparator_(key, *pos);
+      if (replacing_entry && value == pos->second) {
+        return *this;
+      }
+    }
+
+    // Copy the segment before the found position. If not found, this is
+    // everything.
+    auto copy = std::make_shared<array_type>();
```

Comment:
Optional: you can use the new constructor here to avoid the append on the very next line.

Reply:
Done.